Don't use Targets* helper properties in libs

ViktorHofer opened the pull request:
Depends on https://github.com/dotnet/runtime/pull/64610, https://github.com/dotnet/arcade/pull/8459

### High level goals

1. Allow libraries projects to efficiently/correctly target a single target framework. Today it's not possible for projects to use 
`TargetFramework` over `TargetFrameworks` even when only a single target framework is used: https://github.com/dotnet/runtime/issues/47155. Because of that, we add ~250 project outer-build evaluations (100ms in average per outer build) to the overall build. These evaluations impact both incremental restore, incremental build and the overall build.
2. It makes dotnet/runtime libraries use the same common path as our customers when a project needs to condition on platforms.
3. Tooling works best when we don't rely on global properties being set. Today, the `Targets*` properties rely on the `TargetFramework` property being passed in globally during the dispatch from an outer to the inner builds. During design time builds and eventually in other scenarios, the TargetFramework property isn't passed in globally and instead is dynamically injected after the first project's property group. This change will avoid the dependency on the global property and will better support tooling like Visual Studio and Visual Studio Code. Nasty hacks like https://github.com/dotnet/runtime/blob/04bb6eb3f182c934466af9c6f3d5a9ce8efdf716/eng/BeforeTargetFrameworkInference.targets#L3 can be removed entirely.

### More detailed summary

This change makes it possible to migrate 200+ (ref+src) projects to use
TargetFramework instead of TargetFrameworks which avoids the additional
outer build evaluation and invocation which ultimately makes the overall
build faster. As a side effect, this removes yet another artificial difference
between the libraries projects and customer projects. The conditions being
used in our projects are mostly `TargetFrameworkIdentifier`,
`TargetPlatformIdentifier` and `IsTargetFrameworkCompatible`. As a .NET
developer I would use exactly the same conditions in my own project.

Targets* properties (i.e. TargetsWindows, TargetsAnyOS, TargetsUnix,
etc.) rely on the TargetFramework property which usually are set
inside a project. The TargetFramework property is only
available before a project specifies it if it's explicitly set in a props
file or if the project is cross-targeting and the outer-build dispatches
into the inner-build. During the dispatch, the TargetFramework property
is passed in as a global property.
    
Until now that behavior wasn't a problem because every libraries project
cross-targeted (by setting the TargetFrameworks property) even though
many only include a single TargetFramework (i.e. NetCoreAppCurrent).
 
To allow projects to use the TargetFramework property instead of
TargetFrameworks, the Targets* helper properties can't be calculated
anymore early in a props file as the TargetFramework property isn't set
at that time.
    
In general, the guidance by the SDK/msbuild team is to not read from the
TargetFramework property before the project sets it
(in a property group). That effectively means that the TargetFramework
property shouldn't be used in props files at all.
   
Therefore these helper properties can't be used anymore for property
conditions and I'm replacing their usage with TargetPlatformIdentifier
comparisons for both properties and items.
In nearly all cases, the Targets* helper properties can be replaced with
TargetPlatformIdentifier checks on items and in the few cases where
TargetsUnix or TargetsLinux marks multiple tfms as compatible, the exact
tfms must be used instead for the TargetPlatformIdentifier comparison.

Whenever a project needs to condition properties on the platform, I'm
first setting the TargetPlatformIdentifier the same way the SDK sets it
so that the SDK later doesn't need to set it again to avoid the
additional expensive msbuild function call.

am11 commented on src/libraries/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj:
```diff
@@ -1,12 +1,16 @@
+  <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
+  <PropertyGroup>
+    <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
+    <DefineConstants Condition="'$(TargetPlatformIdentifier)' == 'windows'">$(DefineConstants);TargetsWindows</DefineConstants>
```
Existing issue (#39174), but it would be a nice cleanup if we consolidate `TARGET_<platform/architecture>` constants.
In most places, we use `;TARGET_WINDOWS` constant, then in some places we use `;TargetsWindows`.

Perhaps we can move https://github.com/dotnet/runtime/blob/ce15dc7d1416ce8ebc4cffe7ea8c51b071f2fd3c/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L30-L41 and https://github.com/dotnet/runtime/blob/ce15dc7d1416ce8ebc4cffe7ea8c51b071f2fd3c/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L30-L41 into a top-level props and then consistently use them in managed code throughout the repo.

ViktorHofer replied:
Agreed that consolidating constants would be a good idea but I wouldn't want to do that as part of this PR. I have very limited time atm (I'm still on holiday) and want to focus on a few specific goals.

The linked issue doesn't seem to complex. Maybe the SDK already offers defines for platforms?

am11 replied:
Yup, it's an existing issue, not for this PR. Just wanted to highlight it since PR is touching this area. :)

> Maybe the SDK already offers defines for platforms?

When new platforms are added, it's easier to add an entry in the SDK hook (`Targets<NewPlatformName>` property in `eng/BeforeTargetFrameworkInference.targets `) in runtime repo, before full SDK support is available. I was thinking that we could define `TARGET_` constant, in the same vein, in top level props to avoid updating proj files individually. That constant will be unused in most projects (as generally partial classes are preferred over `#if TARGET_`), but most constants defined by SDK are also unused, so it's just good for consistency.

ViktorHofer replied:
The SDK already sets platform provides platform defines and it doesn't differentiate between well known platforms and new/unknown ones: https://github.com/dotnet/sdk/blob/997376c6bc798af22633a8eb36074a0acec8a200/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets#L193-L204.

It sets the define based of the TargetPlatformIdentifier and TargetPlatformVersion, which we provide for the custom, unknown platforms. The only thing that currently can't be expressed is inheritance, as the SDK doesn't know that `#if UNIX` is an applicable constant not just for the `UNIX` TargetPlatformIdentifier but also for i.e. the `LINUX` TargetPlatformIdentifier. That said, expressing inheritance isn't the common case, often it's a direct match i.e. `net7.0-android` -> `#if ANDROID`.

I would start with cleaning up all custom platform defines which aren't part to something, i.e. `Android`, `iOS`, `tvOS`, etc.

ericstj commented:
Similar to your other PR, we need to document what our rules are and if possible try to catch incorrect use.  I'm supportive of the restriction here, but it will be a very subtle distinction that you can use `Targets*` in an `ItemGroup` but not `PropertyGroup`.  I can imagine folks introducing bugs that might slip through PR or at least cause a significant amount of head-scratching.

ViktorHofer commented:
> I'm supportive of the restriction here, but it will be a very subtle distinction that you can use Targets* in an ItemGroup but not PropertyGroup. I can imagine folks introducing bugs that might slip through PR or at least cause a significant amount of head-scratching.

Agreed and I'm a bit worried about this as well. I can definitely add/update documentation to point out this restriction but I doubt this is something that can be protected/validated without writing an msbuild formatter/analyzer.

Do you have any more ideas how to best roll this change out? Ideally all platform conditions, regardless of where they are applied on (property, item, import) would use the `TargetFrameworkIdentifier` property. Unfortunately I don't think that is possible in all cases today as NuGet doesn't understand compatibility between platforms/RIDs today.
Would it make sense to switch existing item conditions that map 1:1 to platforms like Android, iOS, tvOS, MacCatalyst, Linux and Windows as well over to use `TargetPlatformIdentifier` and only keep the few `Targets*` helper properties that are used for inheritance checks like `TargetsUnix`?

ericstj commented:
> without writing an msbuild formatter/analyzer.

Might be worth it at some point if we have enough conventions / style to justify it. 

> Ideally all platform conditions, regardless of where they are applied on (property, item, import) would use the TargetFrameworkIdentifier property

Agreed.  At first this is what I thought you were doing, but then I realized you were only doing PropertyGroups

> inheritance checks like TargetsUnix

I wonder how many of these we have?  If it's small enough we could just say remove them all and make the few cases that use `TargetsUnix` slightly more complicated by checking all the values that are actually targeted.

> Do you have any more ideas

I could imagine all our projects have some boilerplate that parses the platform.  Right now you've copy-pasted that boilerplate (`TargetPlatformIdentifier`).  I could imagine we put that boilerplate in props file and instead insert an import.  That props file could do whatever parsing we want.  We could have a target that makes sure that import is present whenever a project has platform-specific TargetFramework values.  I'm not sure if I like it or hate it, but it does define a process that is self-validating.

ViktorHofer committed:
Don't use Targets* helper properties in libs projs

This change makes it possible to migrate 200+ (ref+src) projects to use
TargetFramework instead of TargetFrameworks which avoids the additional
outer build evaluation and invocation which ultimately makes the overall
build faster.

Targets* properties (i.e. TargetsWindows, TargetsAnyOS, TargetsUnix,
etc.) rely on the TargetFramework property which usually are set
inside a project. The TargetFramework property is only
available before a project specifies it if it's explicitly set in a props
file or if the project is cross-targeting and the outer-build dispatches
into the inner-build. During the dispatch, the TargetFramework property
is passed in as a global property.

Until now that behavior wasn't a problem because every libraries project
cross-targeted (by setting the TargetFrameworks property) even though
many only include a single TargetFramework (i.e. NetCoreAppCurrent).

To allow projects to use the TargetFramework property instead of
TargetFrameworks, the Targets* helper properties can't be calculated
anymore early in a props file as the TargetFramework property isn't set
at that time.

In general, the guidance by the SDK/msbuild team is to not read from the
TargetFramework property before the project sets it
(in a property group). That effectively means that the TargetFramework
property shouldn't be used in props files at all.

Therefore these helper properties can't be used anymore for property
conditions and I'm replacing their usage with TargetPlatformIdentifier
comparisons for both properties and items.
In nearly all cases, the Targets* helper properties can be replaced with
TargetPlatformIdentifier checks on items and in the few cases where
TargetsUnix or TargetsLinux marks multiple tfms as compatible, the exact
tfms must be used instead for the TargetPlatformIdentifier comparison.

Whenever a project needs to condition properties on the platform, I'm
first setting the TargetPlatformIdentifier the same way the SDK sets it
so that the SDK later doesn't need to set it again to avoid the
additional expensive msbuild function call.

ViktorHofer commented:
I believe this change is now ready as well but let me do some additional validation and verification tomorrow. Also I'm waiting for the Helix test results to see if I got something wrong.

ViktorHofer commented:
@ericstj @safern this change is now ready and I'm pretty happy about the outcome. The `TargetPlatformIdentifier` property can be used condition based on the platform (see the updated documentation). For .NETFramework and .NETStandard it returns an empty string (no platform).

The change itself is very isolated but I would still prefer it if you could send out a quick notice to the libraries team announcing it before it gets merged. It most probably is just a matter of pointing to the updated docs file.

I'm still waiting for the Helix tests to make sure I didn't regress any functionality. Maybe you can already review the PR and if everything looks good approve it but not merge in before tomorrow, after a quick notice is sent to the team? Does that sound good?

ViktorHofer commented:
/azp run runtime

ericstj commented on docs/coding-guidelines/project-guidelines.md:
````diff
@@ -80,37 +76,74 @@ When building an individual project the `BuildTargetFramework` and `TargetOS` wi
+<ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(TargetPlatformIdentifier)' == 'windows'">...</ItemGroup>
 ```
-3. Use a StartsWith if the TargetFramework is overloaded with the OS portion.
+5. Set the `TargetPlatformIdentifier` property in the project to be able to condition on it in properties in the project file.
````
I was thinking about this and wondering if we could get away with calculating this in a props if we guarantee the following:
1. Whenever a framework with a TargetPlatformIdentifer exists, there must be a framework without a TargetPlatformIdentifier.  We need this for P2Ps to work well, and most of the times it's good practice to ensure we have a default implementation.  We could enforce this with a target.
2. 1 ensures that any project with a TargetPlatformIdentifier will have more than one TargetFramework, and must have an outer build.
3. Therefore, we can assume that whenever a TargetPlatformIdentifier is present in the TargetFramework it will be passed as a global property, and we can compute the TargetPlatformIdentifier in props.

ericstj replied:
We might be able to even get away without forcing 1, if we just caught the one case where it doesn't work, and forced that case to define it in the project.

We can observe the value of TargetPlatformIdentifier at the start of our Directory.Build.targets evaluation, so we can see what its value is for any of our projects and capture that early on in property evaluation.  We can then see what the value is after the SDK computes it later.  Whenever they mismatch we can raise an error and tell the developer to add a TargetPlatformEvaluation into the project.

ericstj replied:
Nevermind, now I remember why we cannot garuntee this --> design time build.  It never passes TargetFramework values as a global property.  It still sets them with "SetProperty" which effectively appends to the first PropertyGroup.  That's a real bummer.

If we were to get that fixed I think we could clean up this boilerplate.

ViktorHofer replied:
I was trying to avoid any TargetFramework reads or write before the project file itself is evaluated even if the TargetFramework property would be passed in as a global property mostly because other components like the SDK don't do that either. Even though it's boilerplate, it's WYSIWYG so I'm ok with it.

ViktorHofer commented:
@ericstj @safern any additional feedback? As mentioned it would be great if we could also send out a notice to the team as heads-up.

ericstj reviewed (approved):
I gave it another scan and it seems OK to me.  I would appreciate another set of eyes as well.
We might want to wait for Monday to merge this and send the mail.  @safern -- any thoughts on timing?

joperezr commented on src/libraries/Microsoft.VisualBasic.Core/src/Microsoft.VisualBasic.Core.vbproj:
```diff
@@ -9,19 +9,21 @@
+  <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
-    <ILLinkDescriptorsXml Condition="'$(TargetsWindows)' == 'true'">$(MSBuildProjectDirectory)\ILLink\ILLink.Descriptors.Windows.xml</ILLinkDescriptorsXml>
+    <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
```
Instead of duplicating this property everywhere it is used, should we just instead move it to a parent Directory.Build.Props so it is defined already for projects similar to the way TargetsWindows used to?

ViktorHofer replied:
That property relies on `TargetFramework` which isn't available in non cross-targeting builds or during design build. See https://github.com/dotnet/runtime/pull/64500#discussion_r799621575. That means we read from the TargetFramework property in props files.

ViktorHofer commented:
> We might want to wait for Monday to merge this and send the mail. @safern -- any thoughts on timing?

Feel free to merge at your convenience, the PR is ready. For communication I would mention that for src/libraries projects, the `TargetPlatformIdentifier` property replaces most of the `Targets*` properties (except the few that are defined centrally, like `TargetsMobile` which aren't inferred from the `TargetFramework` but from the `TargetOS`. What I would also mention is that `TPI` conditions are direct matches only, i.e. `TPI == 'Unix'` doesn't include `Android` and other child platforms.

ViktorHofer commented:
Ping @ericstj and @safern

ericstj commented:
Addressed the merge conflict.  We'll try to get this in this week, I know folks are busy with some codeflow tasks right now.

ViktorHofer commented:
Thanks for the update Eric, though can you please tell me a more precise target merge date? Asking as the PR was ready last Friday and now after merge conflicts got resolved, is failing again.

Also I used Octokit to see which current open PRs could be impacted by this change and here's the result. Most of them are false positives and could be easily resolved but a few will need updates:
- https://github.com/dotnet/runtime/pull/64868
- https://github.com/dotnet/runtime/pull/64834
- https://github.com/dotnet/runtime/pull/64623
- https://github.com/dotnet/runtime/pull/64500
- https://github.com/dotnet/runtime/pull/64493
- https://github.com/dotnet/runtime/pull/64279
- https://github.com/dotnet/runtime/pull/64062
- https://github.com/dotnet/runtime/pull/63998
- https://github.com/dotnet/runtime/pull/47640

safern commented:
I'm not opposed to this change as this improves build time and also the confusion around what should I use when I have a single target framework in a project within the dotnet/runtime repository. The only concern that I have is that the only validation to avoid usage of these properties will be either a wrong compilation (which might drive people nuts to not understand why the right code is not compiled) or PR Review... but I guess it is fine and people can get use to using `TargetPlatformIdentifier`. I'll get an email prepared so that we can merge today and we also need to fix the current build failures.

ViktorHofer commented:
> I'm not opposed to this change as this improves build time and also the confusion around what should I use when I have a single target framework in a project within the dotnet/runtime repository. The only concern that I have is that the only validation to avoid usage of these properties will be either a wrong compilation (which might drive people nuts to not understand why the right code is not compiled) or PR Review... but I guess it is fine and people can get use to using TargetPlatformIdentifier.

Correct. Such a change will impact developers and it needs clear documentation and communication. This change makes sure all usages of the `Targets*` properties are replaced correctly with `TPI` conditions. The motivation for this change is:
1. Allow libraries projects to efficiently/correctly target a single target framework. Today it's not possible for projects to use 
`TargetFramework` over `TargetFrameworks` even when only a single target framework is used: https://github.com/dotnet/runtime/issues/47155. Because of that, we add ~250 project outer-build evaluations (100ms in average per outer build) to the overall build. These evaluations impact both incremental restore, incremental build and the overall build.
2. It makes dotnet/runtime libraries use the same common path as our customers when a project needs to condition on platforms.
3. Tooling works best when we don't rely on global properties being set. Today, the `Targets*` properties rely on the `TargetFramework` property being passed in globally during the dispatch from an outer to the inner builds. During design time builds and eventually in other scenarios, the TargetFramework property isn't passed in globally and instead is dynamically injected after the first project's property group. This change will avoid the dependency on the global property and will better support tooling like Visual Studio and Visual Studio Code. Nasty hacks like https://github.com/dotnet/runtime/blob/04bb6eb3f182c934466af9c6f3d5a9ce8efdf716/eng/BeforeTargetFrameworkInference.targets#L3 can be removed entirely.

> Also it is a bummer that to condition for properties we have to now use [MSBuild]::GetTargetPlatformIdentifier to get the TPI value, but for items we can directly access the TPI property. I think while this brings value, this will bring some confusion to the table as the Targets* properties are/where simpler as you didn't have to think much about when/where to use them vs now if I want to use it to condition a property I need to do a "more" complicated condition by calling an MSBuild function to get the TPI value.

The idea is that properties use `TPI` to condition on specific platforms (see PR changes) but whenever that's the case, a boilerplate line that initially sets the `TPI` property must be present. It's boilerplate, yes but I don't think it can be avoided because of msbuild's evaluation order. I don't see any inherent problem with that requirement as long as it's well documented and communicated: 
_"Whenever you want to condition properties on platforms, make sure that the `TargetPlatformIdentifier` property is set (see other projects or docs page)."_

> I'll get an email prepared so that we can merge today but we also need to fix the current build failures.

I was already asleep when you responded. Maybe today? :)

Btw here's the code that logs the PRs that are impacted by this change (including a few false positives which have changes in adjacent lines): https://gist.github.com/ViktorHofer/a69aa2cd8e1bddbfe3bee199a42d681c.

ViktorHofer commented:
@safern @ericstj I again resolved merge conflicts and fixed the failure that was happening due to a recent commit in main in which the Targets* property was used. Would really appreciate it if you could let me know what's left for this PR to get merged 🙏 Thank you

stephentoub commented:
I haven't read through the comments here yet, but just taking a look as someone who spends a fair amount of time in the csproj files...

At a high-level, the change seems ok.  Some csprojs will get an extra few lines they don't already have, and conditions like `Condition=" '$(TargetsOSX)' == 'true'"` change to conditions like `Condition=" '$(TargetPlatformIdentifier)' == 'OSX'"`.  Ok.

I'm more concerned that some such conditions are more confusing and much more likely to lead to user error that's hard for us to understand and debug, and we'll end up with devs spinning their wheels trying to figure out what's gone wrong.  For example, one line changes this very obvious statement:
`<ItemGroup Condition=" '$(TargetsUnix)' == 'true'">`
which is pretty obvious in meaning, to:
`<ItemGroup Condition="'$(TargetPlatformIdentifier)' != '' and '$(TargetPlatformIdentifier)' != 'windows'">`
which, at least to me, is not at all obvious.  I fully expect folks will get that conversion wrong.  And there's obviously a subtly I don't understand, because elsewhere this:
`Condition="'$(TargetsUnix)' == 'true' or '$(TargetsBrowser)' == 'true'"`
is replaced by this:
`Condition="'$(TargetPlatformIdentifier)' == 'Unix' or '$(TargetPlatformIdentifier)' == 'Browser'"`
Why is Unix usable in one TargetsUnix case and not the other?  Why are we checking for empty identifiers?  Etc.  To make this usable, that will need to be simplified.

Changes like this one are also harder to understand (at least for me) without digging in further to what's going on:
Before: `<ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' or '$(TargetsWindows)' == 'true'">`
After: `<ItemGroup Condition="'$(GeneratePlatformNotSupportedAssemblyMessage)' == ''">`

Separately, I'm not clear on changes like this one, which is removing a netstandard2.1 build apparently?
https://github.com/dotnet/runtime/pull/64500/files#diff-45edf6f0e5edd19513c971472903d2723a2d094773d9122b8726a684556c0248L4

The motivation called out at the very beginning of the PR states:
"This change makes it possible to migrate 200+ (ref+src) projects to use
TargetFramework instead of TargetFrameworks which avoids the additional
outer build evaluation and invocation which ultimately makes the overall
build faster."
but that change doesn't appear to be in this PR.  If we're going to change everything over and make everyone on the team learn new more complicated conditions, it should come with a benefit... that benefit would seem to require changing those projects over to use TargetFramework, so it'd be nice if that were included in this PR.  A separate commit is fine.

joperezr commented:
You are right that this new system does require a bit more knowledge of how things work so for somebody that is not very familiar with that this will introduce confusion. As to your specific question regarding `<ItemGroup Condition="'$(TargetPlatformIdentifier)' != '' and '$(TargetPlatformIdentifier)' != 'windows'">`, the reason why we now have to check for an empty identifier will mainly be depending on the TargetFrameworks that the specific project is targeting. 

For example, let's take System.Diagnostic.Process.csproj which is one of the ones that recieved the exact change that you pointed out. It's target frameworks are: `<TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-FreeBSD;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-OSX;$(NetCoreAppCurrent)-MacCatalyst;$(NetCoreAppCurrent)-iOS;$(NetCoreAppCurrent)-tvOS;$(NetCoreAppCurrent)</TargetFrameworks>` Previously we had an _inclusive condition_ (`TargetsUnix == true`) because out of all of those TFMs, TargetsUnix was automatically being set to true for all of the TFMs where it was targeting a specific platform that was not windows, so the condition was simple. This PR is removing the TargetsUnix property, so we now want to do an _exclusive condition_ and single out the case for the two TFMs where we don't want this to apply: `$(NetCoreAppCurrent)-windows` and `$(NetCoreAppCurrent`.

All that said, I totally understand your sentiment where even though it is easy to understand for somebody that has the knowledge of the infrastructure and how it works, it may not be so obvious for somebody who isn't that familiar, so this is a fair point to make. If it helps at all, there are very few projects that do all of this multitargeting, and there are also few cases where people changing something in the library have to introduce new conditions like this on a .csproj so I would hope that this shouldn't be a very common problem, but it is a problem nonetheless.

joperezr commented:
That could be an option, yes. That said, it is not uncommon for a project that multi-targets to use the Targets* properties both for properties and for conditionally compiling items, so I suppose in those cases it would be weird to have different conditions for properties and items. That said it would be interesting to hear what @ViktorHofer thinks about that.

Also, to be fair and something I didn't comment above, is that the fact that you need to look into the `TargetFrameworks` property when conditionally compiling files is not really new, nor is it introduced by these changes. I do agree that it is a bit different the way it is addressed, but in the past you also had this issue. For example, if you had a project that targeted: `net7.0` and `net7.0-windows` where your "unix" build is effectively the platform-independent configuration, you already wouldn't be able to use conditions like:
```xml
<ItemGroup Condition="'$(TargetsUnix)' == 'true'">
   <!-- Add my unix compile items here -->
```

Since your Unix build was actually your AnyOS build, so you would actually need to instead use: `'$(TargetsAnyOS)' == 'true'`

ViktorHofer commented:
> How bad would it be if we moved these "helper" properties into a .targets file, and made the rule "You can't use Targets* properties in property evaluation"? And then we based these "helper" properties off of TargetFramework/TargetPlatformIdentifier/etc. This would be symmetrical to TargetFrameworkIdentifier today - you can't use that property in property evaluation for the same reason.

Initially I just switched properties over to use TPI instead of Targets* properties. Eric then later commented that it is confusing to have TPI conditions for properties and Targets* conditions for items and I agree with him on that. In addition to that I think it's valuable to use the same concept for platform conditions that customers follow and that we document: https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2022#msbuild-targetframework-and-targetplatform-functions. Before .NET 5, TPI wasn't suitable to use as target frameworks didn't support encoding platforms in them and the Targets* helper properties had to be used.

> Then, we could still have "readable" conditions in most places, which @stephentoub is asking for. And if a project does need to condition properties based on TargetPlatformIdentifier, we do it like you have the proposed changes here.

What we are observing here is a limitation that exists in the product and should be fixed there. MSBuild/NuGet doesn't support checking compatibility between platforms yet. Example: `$([MSBuild]::TargetPlatformCompatible('$(TargetFramework)', 'unix'))`

> so it'd be nice if that were included in this PR. A separate commit is fine.

Of course. I wanted to keep the PR "simple" but let me add the change to better demonstrate the pros of this.

> Separately, I'm not clear on changes like this one, which is removing a netstandard2.1 build apparently?

@stephentoub that was an unrelated clean-up. The `netstandard2.1` tfm is unnecessary as both the netstandard2.0 and the netstandard2.1 builds are PNSEs. If it helps I can extract this change into a separate PR. I agree that it would have helped if I had left a comment to explain this change. Will keep this in mind for the next time.

eerhardt commented:
> What we are observing here is a limitation that exists in the product and should be fixed there. MSBuild/NuGet doesn't support checking compatibility between platforms yet. Example: $([MSBuild]::TargetPlatformCompatible('$(TargetFramework)', 'unix'))

Agreed. But until we get that functionality in the product, there's nothing wrong with us having some "helper" functions to make our code clean/readable.

ViktorHofer committed:
Use TargetFramework singular to avoid outer builds

Use TargetFramework instead of TargetFrameworks property whenever a
projects only targets a single target framework. This avoid unnecessary
outer builds and evaluations and makes the build faster.

ViktorHofer commented:
> Changes like this one are also harder to understand (at least for me) without digging in further to what's going on:
Before: <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' or '$(TargetsWindows)' == 'true'">
After: <ItemGroup Condition="'$(GeneratePlatformNotSupportedAssemblyMessage)' == ''">

@stephentoub I reverted all the PNSE change. I thought those might be easier to read but apparently it's the other way around and I introduced unnecessary complexity.

> Agreed. But until we get that functionality in the product, there's nothing wrong with us having some "helper" functions to make our code clean/readable.

@eerhardt I think mixing platform conditions (TPI + Targets*) is more confusing than applying platform conditions with the TPI property (even though inheritance isn't supported with TPI). Imagine the following cases:

1. A project has platform specific properties and items:
```
<PropertyGroup>
  <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-OSX</TargetFrameworks>
</PropertyGroup>

<!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
<PropertyGroup>
  <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
  <GeneratePlatformNotSupportedMessage Condition="'$(TargetPlatformIdentifier)' == 'windows'">...</GeneratePlatformNotSupportedMessage>
</PropertyGroup>

<ItemGroup Condition="'$(TargetsUnix)' == 'true'">
  <Compile Include="..." />
</ItemGroup>
```

vs.

```
<PropertyGroup>
  <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-OSX</TargetFrameworks>
</PropertyGroup>

<!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
<PropertyGroup>
  <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
  <GeneratePlatformNotSupportedMessage Condition="'$(TargetPlatformIdentifier)' == 'windows'">...</GeneratePlatformNotSupportedMessage>
</PropertyGroup>

<ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'Linux' or '$(TargetPlatformIdentifier)' == 'OSX'">
  <Compile Include="..." />
</ItemGroup>
```

Today only two `Targets*` properties which support inheritance exist (`TargetsUnix` and `TargetsLinux` for Android) and only very few projects depend on that feature. If we think that this change makes the projects that leverage this feature unreadable, we could define such a helper property per project in a property group so that it can be used both for property and item conditions. That said, based on what I saw while making these changes, I don't think that's necessary.

Another fundamental problem with the libraries specific `Targets*` properties is that they have a totally different meaning than the `Targets*` properties that are repo-wide defined, i.e. `TargetsMobile`. The former is inferred from the TargetFramework and hence means the platform that the target framework targets. The latter is inferred from the TargetOS and is just a shorthand form for `' $(TargetOS)' == '...'` conditions. Because of their equal prefix, people used them equally even though they mean something totall different. 
As an example here's one of such mistakes in main today: https://github.com/dotnet/runtime/blob/1c80769c0d6e2241cd8a3eeb17cab306d0954d1f/src/libraries/System.Runtime.Serialization.Formatters/src/System.Runtime.Serialization.Formatters.csproj#L12-L14 This condition is only true when the `TargetOS` is one of the mobile OSs but not when the target framework platform is a mobile one. That means that if you compile this library i.e. on a Windows OS the condition will be false vs. when you compile the library on a mobile OS.

akoeplinger commented on docs/coding-guidelines/project-guidelines.md:
````diff
@@ -37,22 +37,18 @@ Each project will define a set of supported TargetFrameworks
+Non cross-targeting project that targets .NETStandard:
 ```
 <PropertyGroup>
   <TargetFrameworks>netstandard2.0</TargetFrameworks>
````
should this be TargetFramework too now?

akoeplinger commented on docs/coding-guidelines/project-guidelines.md:
````diff
@@ -80,37 +76,64 @@ When building an individual project the `BuildTargetFramework` and `TargetOS` wi
+Example:
+```
+<PropertyGroup>
+  <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(TargetFrameworkMinimum)</TargetFrameworks>
````
NetFrameworkMinimum instead of TargetFrameworkMinimum?

akoeplinger commented on docs/coding-guidelines/project-guidelines.md:
````diff
@@ -80,37 +76,64 @@ When building an individual project the `BuildTargetFramework` and `TargetOS` wi
 ```
 <PropertyGroup>
-  <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
+  <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(TargetFrameworkMinimum)</TargetFrameworks>
````
same here

akoeplinger commented on docs/coding-guidelines/project-guidelines.md:
```diff
@@ -80,37 +76,64 @@ When building an individual project the `BuildTargetFramework` and `TargetOS` wi
 </PropertyGroup>
+<ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">...</ItemGroup>
 <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">...</ItemGroup>
+<ItemGroup Condition="'$(TargetFramework)' == '$(TargetFrameworkMinimum)'">...</ItemGroup>
```
and here

akoeplinger commented on docs/coding-guidelines/project-guidelines.md:
```diff
@@ -80,37 +76,64 @@ When building an individual project the `BuildTargetFramework` and `TargetOS` wi
+</PropertyGroup>
+<!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
+<PropertyGroup>
+  <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
```
could we move this to a .props file that we import here so that we don't need to have it in every project?

ViktorHofer replied:
Unfortunately not as it relies on the TargetFramework property. See https://github.com/dotnet/runtime/pull/64500#discussion_r799879098. @joperezr already asked the same question :)

akoeplinger replied:
@ViktorHofer my suggestion was a little different, I meant 

```
<Import Project="$(RepositoryEngineeringDir)tpi.props"/>
<!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
<PropertyGroup>
   <DefineConstants Condition="'$(TargetPlatformIdentifier)' == 'android'">$(DefineConstants);ANDROID_USE_BUFFER</DefineConstants>
</PropertyGroup>
```

ViktorHofer replied:
You would trade one line to assign the TPI property with one line to import the props file in which you then again find the property assignment. I don't think we would need any additional helper properties.
An import is more expensive than a property assignment and our project evaluations are already unusually high so I would rather not have such an import unless we think that file would be a good place for other additional helper properties that are dependent on the TargetFramework property.

akoeplinger commented:
In general this seems like a good change, I don't think getting rid of the TargetsUnix inheritance thing is a big deal apart from the potential readability concerns that Stephen mentioned.

safern commented:
So it seems like we reached an agreement on taking this change? If so, I will start prepping the email to send out to the contributors to notify them about it and merge it.

eerhardt commented:
> So it seems like we reached an agreement on taking this change? 

No objection from me.

ViktorHofer commented:
@safern The .NET Framework leg is currently failing. The root cause is understood (currently the .NETFramework leg is overbuilding which will be fixed with https://github.com/dotnet/arcade/pull/8459) and I will submit a fix for it probably tomorrow when the arcade update is consumable. Can you meanwhile please send out the mail and let me know here on GH when that's done? (I currently don't read my work mails 🌵) Thanks a lot everybody for the continued support and your feedback.

@stephentoub I agree that your readability concern is an actual issue. Still, as it only affects a small number of projects I think we should live with it for the sake of the improvement that this PR offers, but drive towards a feature in the sdk/msbuild/nuget that eventually will allow us form compatibility conditions based on platforms (as proposed earlier in this thread).

safern commented:
> Can you meanwhile please send out the mail and let me know here on GH when that's done?

Yup, I'll do that now.

ViktorHofer commented:
> Actually it is already late here on a Friday, so I think my email would just be missed by a ton of people, we don't you go ahead and merge over the weekend and I can send the email on Monday morning?

@safern just to close the loop on this, did you already send out the notice for this change?

safern commented:
Thanks for the reminder, I just sent the notice out 😄

ViktorHofer commented:
Thanks for sending the mail out Santi. I just quickly read over it and I fear that it might cause unnecessary confusion and might trigger negative reactions. I don't think me responding to it would be a wise choice as I'm still on leave for the next two months and as you are the sender so I kindly ask you if you would mind sending out a quick follow-up pointing people to the doc:

> As the previous message might have caused some confusion, please see https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/project-guidelines.md#targetframework-conditions which nicely summarizes how to condition on specific target frameworks and platforms.

Thank you.

ViktorHofer commented:
Yes I missed that part in your mail. Then any remaining confusion on this topic should already be covered by the link to the doc. Thanks Santi
